beluga/include/beluga/sensor/data/dense_grid.hpp

```diff
@@ -0,0 +1,162 @@
+ *   y-axis, as an `std::size_t` value.
+ * - given possibly const grid cell coordinates `xi` and `yi` of type `int`,
+ *   `g.contains(xi, yi)` checks whether such cell is included in the grid;
+ * - given possibly const grid cell coordinates `pi` of `Eigen::Vector2i` type,
+ *   `g.contains(p)` checks whether such cell is included in the grid;
+ * - given possibly const grid cell coordinates `xi` and `yi` of type `int`,
+ *   `g.index_at(xi, yi)` retrieves the corresponding cell index of some type;
+ * - given possibly const grid cell coordinates `pi` of `Eigen::Vector2i` type,
+ *   `g.index_at(pi)` retrieves the corresponding cell index of some type;
+ * - given possibly const grid cell coordinates `xi` and `yi` of type `int`,
+ *   `g.data_at(xi, yi)` optionally returns cell data, if cell is included;
+ * - given possibly const grid cell coordinates `pi` of `Eigen::Vector2i` type,
+ *   `g.data_at(p)` optionally returns cell data, if cell is included;
+ * - given possibly const grid cell index `i` of some type, `g.data_at(i)`
+ *   optionally returns cell data, if cell is included;
+ * - given possibly const embedding space coordinates `x` and `y` of type `double`,
+ *   `g.data_near(x, y)` optionally returns cell data, if cell is included;
+ * - given possibly const embedding space coordinates `p` of `Eigen::Vector2d` type,
+ *   `g.data_near(p)` optionally returns cell data, if cell is included;
+ * - given possibly const grid cell coordinates `xi` and `yi` of type `int`,
+ *   `g.neighborhood4(xi, yi)` computes the cell 4-connected neighborhood as
+ *   a range of `Eigen::Vector2i` type;
+ * - given possibly const grid cell coordinates `pi` of `Eigen::Vector2i` type,
+ *   `g.neighborhood4(p)` computes the cell 4-connected neighborhood as a
+ *   range of `Eigen::Vector2i` type.
+ */
+
+/// Dense 2D grid base type.
+/**
+ * When instantiated, it satisfies \ref DenseGrid2Page.
+ *
+ * \tparam Derived Concrete dense grid type. It must define
+ * `Derived::width()`, `Derived::height()`, `Derived::resolution()`,
+ * `Derived::data_at(index)`, and `Derived::index_at(int, int)`
+ * as described in \ref DenseGrid2Page.
+ */
+template <typename Derived>
+class BaseDenseGrid2 : public BaseRegularGrid2<Derived> {
+ public:
+  /// Checks if a cell is included in the grid.
+  /**
+   * \param xi Grid cell x-axis coordinate.
+   * \param yi Grid cell y-axis coordinate.
+   */
+  [[nodiscard]] bool contains(int xi, int yi) const {
+    return xi >= 0 && xi < static_cast<int>(this->self().width()) && yi >= 0 &&
+           yi < static_cast<int>(this->self().height());
+  }
+
+  /// Checks if a cell is included in the grid.
+  /**
+   * \param pi Grid cell coordinates.
+   */
+  [[nodiscard]] bool contains(const Eigen::Vector2i& pi) const { return this->self().contains(pi.x(), pi.y()); }
+
+  [[nodiscard]] std::size_t index_at(const Eigen::Vector2i& pi) const { return this->self().index_at(pi.x(), pi.y()); }
```

Comment:
nit: If the index type has no requirements. Can we change this?
```suggestion
  [[nodiscard]] auto index_at(const Eigen::Vector2i& pi) const { return this->self().index_at(pi.x(), pi.y()); }
```

Reply:
Good catch! It no longer applies though, see https://github.com/Ekumen-OS/beluga/pull/188/commits/771ac50e97960cab31f16450833a2b23f58167fc.